lib/bolt/pal.rb

```diff
@@ -607,7 +607,36 @@ def run_task(task_name, targets, params, executor, inventory, description = nil)
       end
     end
 
-    def run_plan(plan_name, params, executor = nil, inventory = nil, pdb_client = nil, applicator = nil)
+    def run_plan(plan_name, params, executor, inventory = nil, pdb_client = nil, applicator = nil)
+      # Start the round robin inside the plan compiler, so that
+      # backgrounded tasks can finish once the main plan exits
+      in_plan_compiler(executor, inventory, pdb_client, applicator) do |compiler|
+        # Create a Fiber for the main plan. This will be run along with any
+        # other Fibers created during the plan run in the round_robin, with the
+        # main plan always taking precedence in being resumed.
+        fiber = Fiber.new do
+          r = compiler.call_function('run_plan', plan_name, params.merge('_bolt_api_call' => true))
+          Bolt::PlanResult.from_pcore(r, 'success')
+        rescue Bolt::Error => e
+          Bolt::PlanResult.new(e, 'failure')
+        end
+
+        # Round robin until all Fibers, including the main plan, have finished.
+        # This will stay alive until backgrounded tasks have finished.
+        future = Bolt::PlanFuture.new(fiber, 0, "Main plan #{plan_name}")
+        @logger.trace("Created main plan future '#{future.name}'. Starting round robin...")
+
+        executor.add_future(future)
+        executor.round_robin until executor.plan_complete?
+
+        # Return the result from the main plan
+        future.value
+      end
+    rescue Bolt::Error => e
+      Bolt::PlanResult.new(e, 'failure')
+    end
+
+    def boltspec_run_plan(plan_name, params, executor = nil, inventory = nil, pdb_client = nil, applicator = nil)
```

Comment:
Hm, initially this was so that the plan itself would also execute serially since it doesn't call `create_future` which in BoltSpec is what executes each step, but now that the main plan does call `create_future` this shouldn't be needed anymore.